contracts/bitizens/AvatarItemToken.sol

```diff
@@ -6,218 +6,280 @@ import "./AvatarItemService.sol";
+  }
+
+  function compareItemSlots(uint256 _tokenId1, uint256 _tokenId2)
+    external
+    view 
+    existToken(_tokenId1)
+    existToken(_tokenId2)
+    returns (bool _res) {
+    require(_tokenId1 != _tokenId2, "compared token shouldn't be the same");
+    Item storage item1 = tokenInfo[_tokenId1];
+    Item storage item2 = tokenInfo[_tokenId2];
+    _res = item1.socket == item2.socket;
+  }
+
+  function getTransferCount(uint256 _tokenId) external view validToken(_tokenId) returns(uint256) {
+    return itemTransferCount[_tokenId];
+  }
+
+  function getMineTrack(uint256 _tokenId) external view validToken(_tokenId) returns (uint256[]) {
+    return mineTrack[_tokenId];
+  }
 
   function getOwnedTokenIds(address _owner) external view onlyOperator returns(uint256[] _tokenIds) {
-    
+    require(_owner != address(0), "address invalid");
     return ownedTokens[_owner];
   }
 
-  function getAvatarItemInfo(uint256 _tokenId)
-    external
+  function getTokenInfo(uint256 _tokenId)
+    external 
     view 
-    returns(
-    uint32 _listNumber,
-    uint16 _setNumber,
-    bool _isBitizenItem, 
-    bool _isSet, 
-    uint8[6] _attr
-    ){
-    (_listNumber,_setNumber,_isBitizenItem,_isSet,_attr) = _getAvatarItemInfo(_tokenId);
+    validToken(_tokenId)
+    returns(address, address, bool, int16, uint256[4] _attr1, uint8[5] _attr2) {
+    Item storage item = tokenInfo[_tokenId];
+    _attr1[0] = item.node;
+    _attr1[1] = item.listNumber;
+    _attr1[2] = item.setNumber;
+    _attr1[3] = item.quality;  
+    _attr2[0] = item.rarity;
+    _attr2[1] = item.socket;
+    _attr2[2] = item.gender;
+    _attr2[3] = item.energy;
+    _attr2[4] = item.ext;
+    return (
+      item.foundedBy,
+      item.createdBy,
+      item.isBitizenItem,
+      item.probability,
+      _attr1,
+      _attr2
+      );
   }
 
-  function getBurnedTokenInfo(uint256 _tokenId) external view returns(  
-    uint32 _listNumber,
-    uint16 _setNumber,
-    bool _isBitizenItem, 
-    bool _isSet, 
-    uint8[6] _attr){
-    require(isBurnByTokenId[_tokenId]);
-    AvatarItem storage item = burnTokens[_tokenId];
-    _listNumber = item.listNumber;
-    _setNumber = item.setNumber;
-    _isBitizenItem = item.isBitizenItem;
-    _isSet = item.isSet;
-    _attr[0] = item.types;
-    _attr[1] = item.rarity;
-    _attr[2] = item.socket;      
-    _attr[3] = item.gender;
-    _attr[4] = item.ext1;
-    _attr[5] = item.ext2;  
+  function isBurned(uint256 _tokenId) external view validToken(_tokenId) returns (bool) {
+    return isBurnedToken[_tokenId];
+  }
+
+  function getBurnedTokenCount() external view returns (uint256) {
+    return burnedTokenIds.length;
+  }
+
+  function getBurnedTokenIdByIndex(uint256 _index) external view returns (uint256) {
+    require(_index < burnedTokenIds.length, "out of boundary");
+    return burnedTokenIds[_index];
+  }
+  
+  function getSameItemTokenIds(uint256 _tokenId) external view validToken(_tokenId) returns(uint256[]) {
```

Comment:
Looks like a cool idea, would it be costly to maintain though? What if we have 100,000 users, and 10,000 of them have the same boot. How big will this array be?

Reply:
Let me have a think of  this

Reply:
I should use index to instead of directly return a array.

Reply:
👍